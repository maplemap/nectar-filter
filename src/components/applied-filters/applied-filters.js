import React, {useContext} from 'react';
import {Tag, Button} from 'antd';
import {ContextApp} from '../../reducer';

import './applied-filters.less';

const NONE = '- none -';

export const AppliedFilters = () => {
    const {state: {appliedFilters}, dispatch} = useContext(ContextApp);
    const appliedFilterTypes = Object.keys(appliedFilters);
    const isaAppliedFilterTypes = appliedFilterTypes.length > 0;

    const removeFilterItem = ({filterType, filter}) => {
        const newFiltersByFilterType = appliedFilters[filterType].filter(item => item !== filter);
        const newAppliedFilters = {...appliedFilters};
        if (newFiltersByFilterType.length === 0) {
            delete newAppliedFilters[filterType];
        } else {
            newAppliedFilters[filterType] = newFiltersByFilterType;
        }

        dispatch({appliedFilters: newAppliedFilters});
    };

    const clearAppliedFilters = () => {
        dispatch({appliedFilters: {}})
    };

    const getAllFilters = () => (
        appliedFilterTypes.map(filterType => (
            appliedFilters[filterType].map(filter => {
                const onClose = () => removeFilterItem({filterType, filter});
                return (
                    <Tag
                        key={filter}
                        closable
                        className="applied-filters__item"
                        onClose={onClose}
                    >
                        {filter}
                    </Tag>
                )
            })
        ))
    );

    return (
        <div className="applied-filters">
            <h2>Applied Filters:</h2>
            <div className="applied-filters__items">
                {isaAppliedFilterTypes ? getAllFilters() : NONE}
                {isaAppliedFilterTypes && (
                    <Button
                        className="applied-filters__clear"
                        onClick={clearAppliedFilters}
                    >
                        Clear All
                    </Button>
                )}
            </div>
        </div>
    )
};
