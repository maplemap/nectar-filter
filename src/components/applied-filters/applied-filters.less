.applied-filters {
  &__items {
    display: flex;
  }

  &__item {
    display: flex;
    align-items: center;
  }

  &__clear {
    margin-left: .3rem;
  }
}
