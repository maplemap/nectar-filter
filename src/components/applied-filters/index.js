export {AppliedFilters} from './applied-filters';
