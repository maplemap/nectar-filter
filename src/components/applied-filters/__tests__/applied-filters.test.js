import React from 'react';
import {shallow} from 'enzyme';
import {Tag, Button} from 'antd';
import {AppliedFilters, NONE} from '../applied-filters';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn().mockImplementation(() => ({state: {filtersData: {}, appliedFilters: {}}})),
}));

const getComponent = props => (
    <AppliedFilters {...{
        ...props
    }}
    />
);
const getShallow = props => shallow(getComponent(props));

describe('AppliedFilters', () => {
    let getContextData;
    let dispatch;

    beforeAll(() => {
        dispatch = jest.fn();
        getContextData = ({filtersData = {}, appliedFilters = {}, isMobile = false}) => ({
            state: {filtersData, appliedFilters, isMobile},
            dispatch
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('when `appliedFilters` is provided', () => {
        let appliedFilters;
        let wrapper;

        beforeAll(() => {
            appliedFilters = {size: ['1x1', '2x2'], color: ['blue']};
            React.useContext.mockReturnValueOnce(getContextData({appliedFilters}));
            wrapper = getShallow();
        });

        it('should render `Tag` component according the count of keys of `appliedFilters`', () => {
            expect(wrapper.find(Tag).length).toBe(Object.values(appliedFilters).flat(1).length)
        });

        describe('when `removeFilterItem` method was invoked', () => {
            it('should call `dispatch` with updated `appliedFilters`', () => {
                wrapper.find(Tag).at(0).props().onClose();

                expect(dispatch.mock.calls[0][0]).toMatchSnapshot();
            });
        });


        describe('when `clearAllFilters` method was invoked', () => {
            it('should call `dispatch` with updated `appliedFilters`', () => {
                wrapper.find(Button).props().onClick();

                expect(dispatch).toHaveBeenCalledWith({appliedFilters: {}});
            });
        });
    });

    describe('when `appliedFilters` is not provided', () => {
        it('should render `NONE` caption', () => {
            expect(getShallow().find('.applied-filters__items').props().children[0]).toBe(NONE);
        });
    });
});
