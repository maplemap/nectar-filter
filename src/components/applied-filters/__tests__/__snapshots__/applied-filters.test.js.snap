// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppliedFilters when \`appliedFilters\` is provided when \`removeFilterItem\` method was invoked should call \`dispatch\` with updated \`appliedFilters\` 1`] = `
Object {
  "appliedFilters": Object {
    "color": Array [
      "blue",
    ],
    "size": Array [
      "2x2",
    ],
  },
}
`;
